src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ReflectionReader.cs

```diff
@@ -0,0 +1,596 @@
+                throw new NotImplementedException();
+            }
+
+            object resultCollection = ReflectionCreateCollection(collectionContract);
+            context.AddNewObject(resultCollection);
+            context.IncrementItemCount(1);
+
+            if (!ReflectionReadSpecialCollection(xmlReader, context, collectionContract, resultCollection))
+            {
+                bool isReadOnlyCollection = false;
+                resultCollection = ReadCollectionItems(xmlReader, context, collectionItemName, collectionItemNamespace, collectionContract, resultCollection, isReadOnlyCollection);
+            }
+
+            return resultCollection;
+        }
+
+        private CollectionReadItemDelegate GetCollectionReadItemDelegate(CollectionDataContract collectionContract)
+        {
+            CollectionReadItemDelegate collectionReadItemDelegate;
+            if (collectionContract.Kind == CollectionKind.Dictionary || collectionContract.Kind == CollectionKind.GenericDictionary)
+            {
+                collectionReadItemDelegate = (xmlReaderArg, contextArg, collectionContractArg, itemTypeArg, itemNameArg, itemNsArg) =>
+                {
+                    return ReflectionReadDictionaryItem(xmlReaderArg, contextArg, collectionContractArg);
+                };
+            }
+            else
+            {
+                collectionReadItemDelegate = GetReflectionReadValueDelegate(collectionContract.ItemType);
+            }
+
+            return collectionReadItemDelegate;
+        }
+
+        private object ReadCollectionItems(XmlReaderDelegator xmlReader, XmlObjectSerializerReadContext context, XmlDictionaryString collectionItemName, XmlDictionaryString collectionItemNamespace, CollectionDataContract collectionContract, object resultCollection, bool isReadOnlyCollection)
+        {
+            string itemName = GetCollectionContractItemName(collectionContract);
+            string itemNs = GetCollectionContractNamespace(collectionContract);
+            Type itemType = collectionContract.ItemType;
+            CollectionReadItemDelegate collectionReadItemDelegate = GetCollectionReadItemDelegate(collectionContract);
+            MethodInfo getCollectionSetItemDelegateMethod = s_getCollectionSetItemDelegateMethod.MakeGenericMethod(itemType);
+            CollectionSetItemDelegate collectionSetItemDelegate = (CollectionSetItemDelegate)getCollectionSetItemDelegateMethod.Invoke(this, new object[] { collectionContract, resultCollection, isReadOnlyCollection });
+
+            int index = 0;
+            while (true)
+            {
+                if (xmlReader.IsStartElement(collectionItemName, collectionItemNamespace))
+                {
+                    object collectionItem = collectionReadItemDelegate(xmlReader, context, collectionContract, itemType, itemName, itemNs);
+                    resultCollection = collectionSetItemDelegate(resultCollection, collectionItem, index);
+                    index++;
```

Comment:
We need to set the resultCollection because collectionSetItemDelegate may create a new resultCollection.